### ISSUE TITLE
Mitigate potential for too many mark-and-sweeps in a row preventing finalizer execution

### ISSUE THREAD

**Issue by svaarala** [svaarala/duktape#1322]
If all (or almost all) memory allocation call sites triggered by a specific script prohibit finalizer execution, it may take a long time before finalizer execution is allowed. To be an issue the script allocation behavior must be very one-sided; see more discussion in https://github.com/svaarala/duktape/issues/1311#issuecomment-274305167.

Try to come up with a reasonable backstop strategy to prevent this behavior. For example, keep track of successive mark-and-sweep rounds running without finalization. Then, in all places where mark-and-sweep flags are restored to allow finalizer execution, check the count and perform an immediate voluntary mark-and-sweep if the count is too high (say N=3).

**Comment by svaarala** [comment]
This also only affects objects that reside in reference loops, or applications which have reference counting disabled.

**Comment by fatcerberus** [comment]
> This also only affects objects that reside in reference loops, or applications which have reference counting disabled.

Also maybe applications using the debugger.

**Comment by svaarala** [comment]
Right, using the debugger and being mostly paused / stepping for long periods of time with all allocations simultaneously being from call sites preventing finalizers.

**Comment by svaarala** [comment]
In Duktape 2.1 finalizer execution is decoupled from mark-and-sweep. Pending finalizers even for mark-and-sweep queued objects is triggered by any refzero situation (except for NORZ macros). Rescue/free decisions are done by the subsequent mark-and-sweep so there's some delay after the refzero before a rescue/free happens.

But for most practical code this should now be a non-issue.

**Comment by svaarala** [comment]
... with the amendment that there may still be relevant debugger use cases where the issue matters.
